### DISCUSSION TITLE
Fixtures running after errors

### DISCUSSION THREAD

**Discussion by nestorrp587** [behave/behave#914] [Q&A] [answered]
Hello,

Please let me know if this is not the right place to post a question like this.

I am trying to implement bdd tests with behave and I have created 2 fixtures that create very heavy objects (spinning up whole vms) that I run before a feature. If both fixtures run successfully everything seems to work but I am seeing an odd behavior when my first fixture fails to run: This first fixture runs and seems to fail because I see a ERROR-HOOK message. So far so good, but then I see that my second fixture also runs, which I didn't expect. I don't see a reason for the second fixture to run at all since both have the same scope (the feature) and if the first fixture failed the test is not going to be run anyway.

Is this expected behavior or am I using fixtures in an incorrect way?

Thank you very much
Nestor

**Comment by jenisys** [comment, marked as answer]
FOLLOWING ASSUMPTIONS (because you described the problem not in enough detail):
* Fixtures are used with `use_fixture()` / `use_fixture_by_tag()` in `before_tag()` hook
* HOOK-ERROR occurs in `before_tag()` hook (in setup part)

Currently, all tags of an entity (Feature, Scenario, ...) cause a call to the `before_tag()` hook. If a failure occurs, this info is internally stored, but all tag-hooks are executed.

POTENTIAL SOLUTIONS (currently):
* Store an attribute in the `context` object and check it in the second fixture (attribute is missing or attribute does not have the expected value/not-none)
* Use a composite-fixture approach (see documentation on fixtures at the end) if this is feasible for you

**Reply by nestorrp587** [reply]
Thank you very much for your answer. All your assumptions were correct and the answer will help me a good deal with my issues.

Regarding the composite fixtures solution, I am not sure I understand their usefulness here: as far as I can see a composite fixture is a fixture that runs other fixtures calling `use_fixture()`. Will failing the first fixture's setup step cause the second to not run if they are called inside a composite fixture?

**Reply by jenisys** [reply]
A composite fixture performs only cleanup for fixtures that are already executed/created.
Therefore, a composite fixture provides the exact functionality that you requested.

NOTES:
* ASSUMPTION: Failure occurs while creating one of its fixtures
* If a failure occurs during the cleanup of one of its fixtures, it continues to ensure that all/most of its fixtures are cleaned up.
